client/libra_wallet/src/mnemonic.rs

```diff
@@ -112,6 +137,94 @@ impl Mnemonic {
     }
 }
 
+/// BitReader reads data from a byte slice at the granularity of a single bit.
+struct BitReader<'a> {
+    bytes: &'a [u8],
+    /// Position from the start of the slice, counted as bits instead of bytes.
+    position: u16,
+}
+
+impl<'a> BitReader<'a> {
+    /// Construct a new BitReader from a byte slice.
+    fn new(bytes: &'a [u8]) -> BitReader<'a> {
+        BitReader { bytes, position: 0 }
+    }
+
+    /// Read the value of 11 bits into a u16.
+    fn read_u11(&mut self) -> u16 {
+        let start_position = self.position;
+        let end_position = start_position + 11;
+        let mut value: u16 = 0;
+
+        for i in start_position..end_position {
+            let byte_index = (i / 8) as usize;
+            let byte = self.bytes[byte_index];
+            let shift = 7 - (i % 8);
+            let bit = u16::from(byte >> shift) & 1;
+            value = (value << 1) | bit;
+        }
+
+        self.position = end_position;
+        value
+    }
+}
+
+struct BitWriter {
+    bytes: Vec<u8>,
+    unused: u16,
+    buffer: u16,
+}
+
+impl BitWriter {
+    /// Create a new `BitWriter` around the given writer.
+    fn new(mnemonic_len: usize) -> BitWriter {
+        BitWriter {
+            bytes: Vec::with_capacity(11 * mnemonic_len / 8 + 1),
+            unused: 8,
+            buffer: 0,
+        }
+    }
+
+    /// Write 11 bits.
+    fn write_u11(&mut self, value: u16) {
```

Comment:
Something tells me this is the ideal candidate for a proptest-based verification of functional correctness, which would also be a good OSS beginner task. Would you mind opening an issue to that effect?